File: minsoo/greedy/gym_suit.py
# https://programmers.co.kr/learn/courses/30/lessons/42862?language=cpp

def solution(n, lost, reserve):
    in_common = set(lost) & set(reserve)
    lost = set(lost) - in_common
    reserve = set(reserve) - in_common

    cannot_attend = set()
    for who in lost:
        if who - 1 in reserve:
            reserve.remove(who - 1)
        elif who + 1 in reserve:
            reserve.remove(who + 1)
        else:
            cannot_attend.add(who)

    return n - len(cannot_attend)

File: minsoo/greedy/gym_suit.cc
// https://programmers.co.kr/learn/courses/30/lessons/42862?language=cpp

#include <string>
#include <vector>

using namespace std;

int solution(int n, vector<int> lost, vector<int> reserve) {
    vector<int> status(n + 1, 0);
    for (const auto &i : lost)
        status[i]--;
    for (const auto &i : reserve)
        status[i]++;
    for (int i = 1; i <= n; i++) {
        if (status[i] == -1) {
            if (status[i - 1] == 1)
                status[i] = status[i - 1] = 0;
            else if (status[i + 1] == 1)
                status[i] = status[i + 1] = 0;
        }
    }
    int attend = 0;
    for (int i = 1; i <= n; i++)
        if (status[i] != -1)
            attend++;
    return attend;
}
